go/libkb/upak_lite.go

```diff
@@ -0,0 +1,359 @@
+package libkb
+
+import (
+	"fmt"
+	"io/ioutil"
+
+	"github.com/buger/jsonparser"
+	keybase1 "github.com/keybase/client/go/protocol/keybase1"
+)
+
+func LoadUPAKLite(arg LoadUserArg) (ret *keybase1.UpkLiteV1AllIncarnations, err error) {
+	uid := arg.uid
+	m := arg.m
+
+	user, err := LoadUserFromServer(m, uid, nil)
+	if err != nil {
+		return nil, err
+	}
+	leaf, err := lookupMerkleLeaf(m, uid, false, nil)
+	if err != nil {
+		return nil, err
+	}
+	loader := NewHighSigChainLoader(m, user, leaf)
+	highChain, err := loader.Load()
+	if err != nil {
+		return nil, err
+	}
+	return buildUpkLiteAllIncarnations(m, user, leaf, highChain)
+}
+
+type HighSigChainLoader struct {
+	MetaContextified
+	user      *User
+	leaf      *MerkleUserLeaf
+	chain     *HighSigChain
+	chainType *ChainType
+	links     ChainLinks
+	ckf       ComputedKeyFamily
+	dirtyTail *MerkleTriple
+}
+
+type HighSigChain struct {
+	MetaContextified
+	uid        keybase1.UID
+	username   NormalizedUsername
+	chainLinks ChainLinks
+	// for a locally delegated key
+	localCki *ComputedKeyInfos
+	// For a user who has never done a reset, this is 1.
+	currentSubchainStart keybase1.Seqno
+	prevSubchains        []ChainLinks
+}
+
+const UpkLiteMinorVersionCurrent = keybase1.UpkLiteMinorVersion_V0
+
+func NewHighSigChainLoader(m MetaContext, user *User, leaf *MerkleUserLeaf) *HighSigChainLoader {
+	hsc := HighSigChain{
+		MetaContextified: NewMetaContextified(m),
+		uid:              user.GetUID(),
+		username:         user.GetNormalizedName(),
+	}
+	loader := HighSigChainLoader{
+		user:             user,
+		leaf:             leaf,
+		chain:            &hsc,
+		chainType:        PublicChain,
+		MetaContextified: NewMetaContextified(m),
+	}
+	loader.ckf.kf = user.GetKeyFamily()
+	return &loader
+}
+
+func (l *HighSigChainLoader) Load() (ret *HighSigChain, err error) {
+	// request new links (and the unverified tail) from the server
+	// and put them into the highSigChain
+	err = l.LoadFromServer()
+	if err != nil {
+		return nil, err
+	}
+	// verify the chain
+	err = l.chain.VerifyChain(l.M())
+	if err != nil {
+		return nil, err
+	}
+	// compute keys
+	err = l.VerifySigsAndComputeKeys()
+	if err != nil {
+		return nil, err
+	}
+	return l.chain, nil
+}
+
+func (l *HighSigChainLoader) selfUID() (uid keybase1.UID) {
+	// for now let's always assume this isn't applicable, because there's
+	// no distinction for loading yourself
+	return uid
+}
+
+func (l *HighSigChainLoader) LoadFromServer() (err error) {
+	srv := l.GetMerkleTriple()
+	l.dirtyTail, err = l.chain.LoadFromServer(l.M(), srv, l.selfUID())
+	return err
+}
+
+func (hsc *HighSigChain) LoadFromServer(m MetaContext, t *MerkleTriple, selfUID keybase1.UID) (dirtyTail *MerkleTriple, err error) {
+	// get the high sigs from the server
+	// ------------------
+	m, tbs := m.WithTimeBuckets()
+
+	apiArg := APIArg{
+		Endpoint:    "sig/get_high",
+		SessionType: APISessionTypeREQUIRED,
```

Comment:
this should be `APISessionTypeOptional` so you can load skipchains when logged out

Reply:
(don't worry about the logged-out test)

Reply:
done. and with a test. :)